### PR TITLE
Spring Boot practice: Add Category model

### PR DESCRIPTION
**Pull request by Chorniypistolet**

### CODE REVIEW

File: src/main/java/mate/academy/spring/boot/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+import java.util.List;
+
+@RestController
+@RequestMapping("/categories")
+@RequiredArgsConstructor
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookMapper bookMapper;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/CategoryController.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    private final BookService bookService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('ADMIN')")
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/CategoryController.java]:
```suggestion
    public CategoryDto createCategory(@RequestBody @Valid CategoryRequestDto categoryDto) {
```

create CategoryRequestDto withou id field

File: src/main/java/mate/academy/spring/boot/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @GetMapping("/{id}")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getById(id);
+    }
+
+    @PutMapping("/{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    public CategoryDto updateCategory(@PathVariable Long id, @RequestBody @Valid CategoryDto categoryDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/CategoryController.java]:
```suggestion
    public CategoryDto updateCategory(@PathVariable Long id, @RequestBody @Valid CategoryRequestDto categoryDto) {
```

File: src/main/java/mate/academy/spring/boot/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id) {
+        return bookService.findAllByCategoryId(id);
+    }
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/CategoryController.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/mapper/CategoryMapper.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+import mate.academy.spring.boot.config.MapperConfig;
+import mate.academy.spring.boot.dto.category.CategoryDto;
+import mate.academy.spring.boot.model.Category;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface CategoryMapper {
+    CategoryDto toDto(Category category);
+    Category toEntity(CategoryDto categoryDTO);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/mapper/CategoryMapper.java]:
```suggestion
    CategoryDto toDto(Category category);
    
    Category toEntity(CategoryDto categoryDto);
```

File: src/main/java/mate/academy/spring/boot/model/Book.java [hunk 6]
```diff
@@ -37,4 +37,12 @@ public class Book {
     private String coverImage;
     @Column(nullable = false, columnDefinition = "TINYINT(1)")
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/spring/boot/model/Book.java [hunk 7]
```diff
@@ -37,4 +37,12 @@ public class Book {
...
+    @JoinTable(
+            name = "book_category",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Book.java]:
```suggestion
            name = "books_categories",
```

File: src/main/java/mate/academy/spring/boot/model/Category.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    private String description;
+    @ManyToMany(mappedBy = "categories", fetch = FetchType.LAZY)
+    private Set<Book> books = new HashSet<>();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Category.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/Category.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    @Column(nullable = false)
+    private boolean deleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Category.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/mate/academy/spring/boot/model/Category.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import org.hibernate.annotations.SQLRestriction;
+import java.util.HashSet;
+import java.util.Set;
+
+@Entity
+@Getter
+@Setter
+@ToString
+@SQLDelete(sql = "UPDATE categories SET deleted = true WHERE id=?")
+@SQLRestriction("deleted = false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Category.java]:
```suggestion
@SQLRestriction("is_deleted = false")
```

File: src/main/java/mate/academy/spring/boot/model/Category.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+import java.util.HashSet;
+import java.util.Set;
+
+@Entity
+@Getter
+@Setter
+@ToString
+@SQLDelete(sql = "UPDATE categories SET deleted = true WHERE id=?")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Category.java]:
```suggestion
@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?")
```

File: src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java [hunk 12]
```diff
@@ -40,9 +41,9 @@ protected void doFilterInternal(HttpServletRequest request,
     }
 
     private String getToken(HttpServletRequest request) {
-        String bearerToken = request.getHeader("Authorization");
-        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
-            return bearerToken.substring(7);
+        String bearerToken = request.getHeader(AUTHORIZATION);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java]:
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

File: src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java [hunk 13]
```diff
@@ -13,15 +13,16 @@
...
 
 @Component
 @RequiredArgsConstructor
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
     private final JwtUtil jwtUtil;
     private final UserDetailsService userDetailsService;
-
+    private static final String AUTHORIZATION = "Authorization";
+    private static final String BEARER = "Bearer ";
+    private static final int INDEX = 7;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java]:
```suggestion
    private static final String TOKEN_HEADER = "Bearer ";
```

File: src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java [hunk 14]
```diff
@@ -40,9 +41,9 @@ protected void doFilterInternal(HttpServletRequest request,
     }
 
     private String getToken(HttpServletRequest request) {
-        String bearerToken = request.getHeader("Authorization");
-        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
-            return bearerToken.substring(7);
+        String bearerToken = request.getHeader(AUTHORIZATION);
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(BEARER)) {
+            return bearerToken.substring(INDEX);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(TOKEN_HEADER.length());
```

File: src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java [hunk 15]
```diff
@@ -40,9 +41,9 @@ protected void doFilterInternal(HttpServletRequest request,
     }
 
     private String getToken(HttpServletRequest request) {
-        String bearerToken = request.getHeader("Authorization");
-        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
-            return bearerToken.substring(7);
+        String bearerToken = request.getHeader(AUTHORIZATION);
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(BEARER)) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/security/JwtAuthenticationFilter.java]:
```suggestion
        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(TOKEN_HEADER)) {
```

File: src/main/java/mate/academy/spring/boot/service/CategoryService.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+import org.springframework.stereotype.Service;
+import java.util.List;
+
+@Service
+public interface CategoryService {
+    List<CategoryDto> findAll();
+    CategoryDto getById(Long id);
+    CategoryDto save(CategoryDto categoryDto);
+    CategoryDto update(Long id, CategoryDto categoryDto);
+    void deleteById(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/CategoryService.java]:
```suggestion
    List<CategoryDto> findAll();
    
    CategoryDto getById(Long id);
    
    CategoryDto save(CategoryDto categoryDto);
    
    CategoryDto update(Long id, CategoryDto categoryDto);
    
    void deleteById(Long id);
```

File: src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public List<CategoryDto> findAll() {
+        return categoryRepository.findAll().stream()
+                .map(categoryMapper::toDto)
+                .collect(Collectors.toList());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java]:
```suggestion
                .toList();
```

File: src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        return categoryRepository.findById(id)
+                .map(categoryMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("No such a category with id: " + id));
```

File: src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toEntity(categoryDto);
+        category = categoryRepository.save(category);
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java]: add id to message

File: src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+        category.setName(categoryDto.getName());
+        category.setDescription(categoryDto.getDescription());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java]:
```suggestion
        categoryMapper.updateCategoryFromDto(categoryDto, category);
```

File: src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+        category = categoryRepository.save(category);
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category not found"));
+        category.setDeleted(true);
+        categoryRepository.save(category);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java]:
```suggestion
        categoryRepository.deleteById(id);
```

File: src/main/resources/db/changelog/changes/09-create-categories-table.yaml [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: deleted
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/09-create-categories-table.yaml]: rename

File: src/main/resources/db/changelog/changes/10-create-book_category-table.yaml [hunk 23]
```diff
@@ -0,0 +1,30 @@
+databaseChangeLog:
+  - changeSet:
+      id: 10
+      author: ChorniyPistolet
+      changes:
+        - createTable:
+            tableName: book_category
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-book_category-table.yaml]: rename

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/spring/boot/repository/book/BookRepository.java [hunk 24]
```diff
@@ -4,9 +4,16 @@
 import mate.academy.spring.boot.model.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
 
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/spring/boot/repository/book/BookRepository.java]:
```suggestion
```

Comment (MentorMateAcademy) [reply src/main/java/mate/academy/spring/boot/repository/book/BookRepository.java]: Not mandatory

File: src/main/java/mate/academy/spring/boot/repository/category/CategoryRepository.java [hunk 25]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.spring.boot.repository.category;
+
+import mate.academy.spring.boot.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/spring/boot/repository/category/CategoryRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/BookService.java [hunk 26]
```diff
@@ -1,10 +1,7 @@
 package mate.academy.spring.boot.service;
 
 import java.util.List;
-import mate.academy.spring.boot.dto.book.CreateBookRequestDto;
-import mate.academy.spring.boot.dto.book.BookDto;
-import mate.academy.spring.boot.dto.book.BookSearchParameters;
-import mate.academy.spring.boot.dto.book.UpdateBookRequestDto;
+import mate.academy.spring.boot.dto.book.*;
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/spring/boot/service/BookService.java]: don`t use * in imports

File: src/main/java/mate/academy/spring/boot/service/impl/BookServiceImpl.java [hunk 27]
```diff
@@ -1,12 +1,8 @@
 package mate.academy.spring.boot.service.impl;
 
 import java.util.List;
-
 import lombok.RequiredArgsConstructor;
-import mate.academy.spring.boot.dto.book.CreateBookRequestDto;
-import mate.academy.spring.boot.dto.book.BookDto;
-import mate.academy.spring.boot.dto.book.BookSearchParameters;
-import mate.academy.spring.boot.dto.book.UpdateBookRequestDto;
+import mate.academy.spring.boot.dto.book.*;
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/spring/boot/service/impl/BookServiceImpl.java]: fix imports

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/spring/boot/service/CategoryService.java [hunk 28]
```diff
@@ -0,0 +1,19 @@
+package mate.academy.spring.boot.service;
+
+import mate.academy.spring.boot.dto.category.CategoryDto;
+import mate.academy.spring.boot.dto.category.CategoryRequestDto;
+import org.springframework.stereotype.Service;
+import java.util.List;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/CategoryService.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+        Category category = categoryMapper.toEntity(categoryRequestDtoDto);
+        category = categoryRepository.save(category);
+        return categoryMapper.toDto(category);
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryRequestDto categoryRequestDto) {
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("No such a category with id: " + id));
+        category = categoryMapper.updateCategoryFromDto(categoryRequestDto, category);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/CategoryServiceImpl.java]:
```suggestion
        categoryMapper.updateCategoryFromDto(categoryRequestDto, category);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**